### DISCUSSION TITLE
Comments added in body

### DISCUSSION THREAD

**Discussion by alexnguyennz** [vercel/next.js#70739] [Help] [answered]
### Summary

I've noticed in a site using Next.js app router (14.2.13), these types of comments are added to the body. Any idea what they're for? They show up in development and production when using `children` in the main layout. Might not actually be related to Next.js but I'm not exactly sure.

```
<!--$-->
<!--/$-->
```

You can also see them in https://vercel.com and https://nextjs.org.

**Comment by suprunchuk** [comment]
The comments you're seeing, `<!--$-->` and `<!--/$-->`, are likely related to the internal workings of React, specifically React Server Components (RSC), which are a feature in Next.js when using the app router. These comments are used by React to demarcate sections of the HTML that correspond to server-rendered components.

**Comment by icyJoseph** [comment, marked as answer]
This is what Suspense boundaries inject. It is used by `react-dom`. No need to worry about them, it has been part of React ever since Suspense was introduced, even before Server Components.

- https://github.com/facebook/react/blob/0751fac747452af8c0494900b4afa7c56ee7b32c/packages/react-dom-bindings/src/server/ReactFizzConfigDOM.js#L3875-L3879

```jsx
// Suspense boundaries are encoded as comments.
const startCompletedSuspenseBoundary = stringToPrecomputedChunk('<!--$-->');
const startPendingSuspenseBoundary1 = stringToPrecomputedChunk(
  '<!--$?--><template id="',
);
```

And a test, where they expect them to be in the HTML:

- https://github.com/facebook/react/blob/0751fac747452af8c0494900b4afa7c56ee7b32c/packages/react-server-dom-fb/src/__tests__/ReactDOMServerFB-test.internal.js#L96

```jsx
    expect(result).toMatchInlineSnapshot(`"<div><!--$-->Done<!--/$--></div>"`);
```

**Comment by alexnguyennz** [comment]
Good to know, thanks!
